tests/e2e/initialization/init_test.go

```diff
@@ -0,0 +1,143 @@
+)
+
+const forkHeight = 10
+
+var (
+	expectedConfigFiles = []string{
+		"app.toml", "config.toml", "genesis.json", "node_key.json", "priv_validator_key.json",
+	}
+)
+
+// TestChainInit tests that chain initialization correctly initializes a full chain
+// and produces the desired output with genesis, chain and validator configs.
+func TestChainInit(t *testing.T) {
+	const id = initialization.ChainAID
+
+	var (
+		nodeConfigs = []*initialization.NodeConfig{
+			{
+				Name:               "0",
+				Pruning:            "default",
+				PruningKeepRecent:  "0",
+				PruningInterval:    "0",
+				SnapshotInterval:   1500,
+				SnapshotKeepRecent: 2,
+				IsValidator:        true,
+			},
+			{
+				Name:               "1",
+				Pruning:            "nothing",
+				PruningKeepRecent:  "0",
+				PruningInterval:    "0",
+				SnapshotInterval:   100,
+				SnapshotKeepRecent: 1,
+				IsValidator:        false,
+			},
+		}
+		dataDir, err = ioutil.TempDir("", "osmosis-e2e-testnet-test")
+	)
+
+	chain, err := initialization.InitChain(id, dataDir, nodeConfigs, time.Second*3, forkHeight)
+	require.NoError(t, err)
+
+	require.Equal(t, chain.ChainMeta.DataDir, dataDir)
+	require.Equal(t, chain.ChainMeta.Id, id)
+
+	require.Equal(t, len(nodeConfigs), len(chain.Nodes))
+
+	actualNodes := chain.Nodes
+
+	for i, expectedConfig := range nodeConfigs {
+		actualNode := actualNodes[i]
+
+		validateNode(t, id, dataDir, expectedConfig, actualNode)
+	}
+}
+
+// TestSingleNodeInit tests that node initialization correctly initializes a single node
+// and produces the desired output with genesis, chain and validator config.
+func TestSingleNodeInit(t *testing.T) {
+	const (
+		id = initialization.ChainAID
+	)
+
+	var (
+		existingChainNodeConfigs = []*initialization.NodeConfig{
+			{
+				Name:               "0",
+				Pruning:            "default",
+				PruningKeepRecent:  "0",
+				PruningInterval:    "0",
+				SnapshotInterval:   1500,
+				SnapshotKeepRecent: 2,
+				IsValidator:        true,
+			},
+			{
+				Name:               "1",
+				Pruning:            "nothing",
+				PruningKeepRecent:  "0",
+				PruningInterval:    "0",
+				SnapshotInterval:   100,
+				SnapshotKeepRecent: 1,
+				IsValidator:        true,
+			},
+		}
+		expectedConfig = &initialization.NodeConfig{
+			Name:               "2",
+			Pruning:            "everything",
+			PruningKeepRecent:  "0",
+			PruningInterval:    "0",
+			SnapshotInterval:   100,
+			SnapshotKeepRecent: 1,
+			IsValidator:        false,
+		}
+		dataDir, err = ioutil.TempDir("", "osmosis-e2e-testnet-test")
+	)
+
+	// Setup
+	existingChain, err := initialization.InitChain(id, dataDir, existingChainNodeConfigs, time.Second*3, forkHeight)
+	require.NoError(t, err)
+
+	actualNode, err := initialization.InitSingleNode(existingChain.ChainMeta.Id, dataDir, filepath.Join(existingChain.Nodes[0].ConfigDir, "config", "genesis.json"), expectedConfig, time.Second*3, 3, "testHash", []string{"some server"})
```

Comment:
I assume the next step here is automating pulling a specified height hash from one of the validator's RPCs? Shouldn't we have enough here already to start filling in the RPC server address? IMO we should just fill this in with a single RPC twice ("1.23.45.67:26657,1.23.45.67:26657"), just makes it simpler to debug. Unless you feel we should be testing with two separate RPCs

Reply:
You are right, the next step would be to start actually utilizing this initialization logic for spinning up a state-synching full node. I don't have that completed yet. Before I continue with this, there is one more large refactoring PR to merge.

And thanks for the advice, I will make sure to use the same one twice